### PR TITLE
Web Inspector: Create All Fonts Tab in Font Section

### PR DESCRIPTION
**Pull request by francescorn**
#### e8bc0489964047b20a92da6fe567e8d315d4b3be
<pre>
Web Inspector: Create All Fonts Tab in Font Section
<a href="https://bugs.webkit.org/show_bug.cgi?id=260127">https://bugs.webkit.org/show_bug.cgi?id=260127</a>

Feature proposal link: <a href="https://docs.webkit.org/Other/Contributor%20Meetings/Slides2023/All%20Fonts%20Feature.pdf">https://docs.webkit.org/Other/Contributor%20Meetings/Slides2023/All%20Fonts%20Feature.pdf</a>

Reviewed by NOBODY (OOPS!).

Create function in DOMFontStyles to retrieve all nodes and child nodes in the DOM.
Create function loadAllStyleSheets in CSSManager to retrieve all fonts in a set in the DOM.
Create allFontsRow detailsSectionRow, allFontsGroup detailsSectionGroup, and allFontsSection detailsSection.
Create function to update allFontsRow when DOMContent is loaded to trigger function that retrieves all fonts in DOM.
Create function _handleAllFontsValuesChanged that retrieves all fonts and calls loadAllStyleSheets and creates an array.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
(WI.CSSManager.prototype.loadAllStyleSheets):
* Source/WebInspectorUI/UserInterface/Models/DOMFontStyles.js: Added.
(WI.DOMFontStyles):
(WI.DOMFontStyles.parseSelectorListPayload):
(WI.DOMFontStyles.prototype.createSourceCodeLocation):
(WI.DOMFontStyles.uniqueOrderedStyles):
(WI.DOMFontStyles.prototype.get node):
(WI.DOMFontStyles.prototype.get matchedRules):
(WI.DOMFontStyles.prototype.get inheritedRules):
(WI.DOMFontStyles.prototype.get inlineStyle):
(WI.DOMFontStyles.prototype.get attributesStyle):
(WI.DOMFontStyles.prototype.get pseudoElements):
(WI.DOMFontStyles.prototype.get computedStyle):
(WI.DOMFontStyles.prototype.get orderedStyles):
(WI.DOMFontStyles.prototype.get computedPrimaryFont):
(WI.DOMFontStyles.prototype.get usedCSSVariables):
(WI.DOMFontStyles.prototype.get allCSSVariables):
(WI.DOMFontStyles.prototype.set ignoreNextContentDidChangeForStyleSheet):
(WI.DOMFontStyles.prototype.get needsRefresh):
(WI.DOMFontStyles.prototype.get uniqueOrderedStyles):
(WI.DOMFontStyles.prototype.refreshIfNeeded):
(WI.DOMFontStyles.prototype.refresh.wrap):
(WI.DOMFontStyles.prototype.refresh.fetchedMatchedStyles):
(WI.DOMFontStyles.prototype.refresh.fetchedInlineStyles):
(WI.DOMFontStyles.prototype.refresh.fetchedComputedStyle):
(WI.DOMFontStyles.prototype.refresh.fetchedFontData):
(WI.DOMFontStyles.prototype.refresh):
(WI.DOMFontStyles.prototype.addRule.completed):
(WI.DOMFontStyles.prototype.addRule.styleChanged):
(WI.DOMFontStyles.prototype.addRule.addedRule):
(WI.DOMFontStyles.prototype.addRule.inspectorStyleSheetAvailable):
(WI.DOMFontStyles.prototype.addRule):
(WI.DOMFontStyles.prototype.effectivePropertyForName):
(WI.DOMFontStyles.prototype.mediaQueryResultDidChange):
(WI.DOMFontStyles.prototype.pseudoClassesDidChange):
(WI.DOMFontStyles.prototype.attributeDidChange):
(WI.DOMFontStyles.prototype.changeRuleSelector.ruleSelectorChanged):
(WI.DOMFontStyles.prototype.changeRuleSelector):
(WI.DOMFontStyles.prototype.changeStyleText):
(WI.DOMFontStyles.prototype._parseSourceRangePayload):
(WI.DOMFontStyles.prototype._parseStylePropertyPayload):
(WI.DOMFontStyles.prototype._parseStyleDeclarationPayload):
(WI.DOMFontStyles.prototype._parseRulePayload):
(WI.DOMFontStyles.prototype._markAsNeedsRefresh):
(WI.DOMFontStyles.prototype._handleCSSStyleSheetContentDidChange):
(WI.DOMFontStyles.prototype._updateStyleCascade):
(WI.DOMFontStyles.prototype._collectStylesInCascadeOrder):
(WI.DOMFontStyles.prototype._markOverriddenProperties.isOverriddenByRelatedShorthand):
(WI.DOMFontStyles.prototype._markOverriddenProperties):
(WI.DOMFontStyles.prototype._associateRelatedProperties):
(WI.DOMFontStyles.prototype._collectCSSVariables):
(WI.DOMFontStyles.prototype._isPropertyFoundInMatchingRules):
* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
(WI.FontStyles):
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel):
(WI.FontDetailsPanel.prototype.update):
(WI.FontDetailsPanel.prototype.layout):
(WI.FontDetailsPanel.prototype.initialLayout):
(WI.FontDetailsPanel.prototype.get _fontPropertiesMap):
(WI.FontDetailsPanel.prototype._updateUIWithFontNames):
(WI.FontDetailsPanel.prototype._createDetailsSectionRowForProperty):
(WI.FontDetailsPanel.prototype._formatPropertyValue):
(WI.FontDetailsPanel.prototype._listAllWebpageFonts):
(_handleAllFontsValuesChanged):
</pre>
https://github.com/WebKit/WebKit/commit/e8bc0489964047b20a92da6fe567e8d315d4b3be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17075 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17400 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17882 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18861 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15975 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20674 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17542 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18190 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17278 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17638 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14815 "4 api tests failed or timed out") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19677 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14877 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15506 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22207 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15877 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15673 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20024 "3 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13792 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15418 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19783 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16094 "Built successfully") | | |

### CODE REVIEW

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 1]
```diff
@@ -23,10 +23,8 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
-{
-    constructor(delegate)
-    {
+WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel {
+    constructor(delegate) {
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: we dont make changes like this unless there's something relevant to modify, so please undo this change

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 2]
```diff
@@ -35,18 +33,23 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
         this._fontStyles = null;
         this._fontVariationRowsMap = new Map;
         this._basicPropertyRowsMap = new Map;
+        this._fontPropertyRowsMap = new Map;
         this._basicPropertyNames = ["font-size", "font-style", "font-weight", "font-stretch"];
 
         this._abortController = new AbortController;
-        this._throttledUpdate = new Throttler(() => { this.update() }, 100);
-        this._debouncedUpdate = new Debouncer(() => { this._skipNextUpdate = false; this.update(); });
+        this._throttledUpdate = new Throttler(() => {
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: ditto above

i realize that this doesn't match the expected style of Web Inspector and can understand wanting to fix it, but again we dont make drive-by style changes unless it's related to the overall purpose of the PR

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 3]
```diff
@@ -35,18 +33,23 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+            this.update();
+        });
         this._skipNextUpdate = false;
     }
 
     // Public
 
-    refresh(significantChange)
-    {
+    refresh(significantChange) {
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: ditto above

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 4]
```diff
@@ -57,8 +60,7 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
         this._throttledUpdate.fire();
     }
 
-    update()
-    {
+    update() {
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: ditto above

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 5]
```diff
@@ -220,6 +222,13 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
 
         let fontVariationPropertiesSection = new WI.DetailsSection("font-variation-properties", WI.UIString("Variation Properties", "Variation Properties @ Font Details Sidebar Section", "Section title for font variation properties."), [this._fontVariationsGroup]);
         this.element.appendChild(fontVariationPropertiesSection.element);
+
+        // Font Properties
+        let allFontsRow = new WI.DetailsSectionSimpleRow(WI.UIString("All Fonts", "All Fonts @ Font Details Sidebar Property", "Property title for all fonts used."));
+
+        this._allFontsGroup = new WI.DetailsSectionGroup([allFontsRow]);
+        let allFontsSection = new WI.DetailsSection("font-all-fonts-used", "All Fonts", [this._allFontsGroup]);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:262]: we don't usually do piecemeal changes like this, where one PR adds some structure only for another PR to fill it in.  usually each PR is a self contained feature (or at least enough of one to be considered something substantive/independent).

i am very in favor of showing additional font information, but as-is this PR is not really complete enough to be approved.  once you have something to put inside this element then we can revisit the overall idea :)

**Review by dcrousso — CHANGES_REQUESTED** [review]

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js (hunk 5 above)
Comment (francescorn) [reply Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:262]: Okay thank you for clarifying, converted to draft.

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 6]
```diff
@@ -538,6 +552,28 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+                fontSet.add(name);
+            });
+        });
+
+        let fontList = Array.from(fontSet);
+        fontList.sort();
+
+        return fontList;
+    }
+
```
Comment (rcaliman-apple) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:619]: This won't do what you think it does. 

The scope of this code is Web Inspector's `Main.html` page, not the inspected page. So, at best, this lists all the strings defined in the `font-family` CSS property of any CSS rule in the Web Inspector tool itself.

To solve this sort of task, a client-side solution is insufficient. 

As mentioned by @dcrousso in Slack, there needs to be a way on the backend (C++) to aggregate the fonts actually loaded and used. Simply iterating through `font-family` string values doesn't guarantee that those fonts 1) were loaded or 2) were used. For example, `font-family: "Nothing"` doesn't say that the font file itself was actually loaded as part of a `@font-face` CSS rule.

The task at hand involves a bit of digging through the backend and identifying where that data exists (@litherum might be able to point you in the right direction). Then, collect that data about used fonts and keep it updated when something on the page changes, like a styles recalculation. Then, the necessary `*Agent`, [InspectorInstrumentation.cpp](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/inspector/InspectorInstrumentation.cpp) and protocol files need to be updated to transport that data from the backend to the client (perhaps [CSS.json](https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/inspector/protocol/CSS.json#L339-L349)) . Then, ultimately  render the list of fonts in the Font sidebar panel.

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 7]
```diff
@@ -262,18 +276,18 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+                break;
+            case "font-style":
+                return new WI.DetailsSectionSimpleRow(WI.UIString("Style", "Style @ Font Details Sidebar Property", "Property title for `font-style`."));
+                break;
+            case "font-weight":
+                return new WI.DetailsSectionSimpleRow(WI.UIString("Weight", "Weight @ Font Details Sidebar Property", "Property title for `font-weight` and `wght` variation axis."));
+                break;
+            case "font-stretch":
+                return new WI.DetailsSectionSimpleRow(WI.UIString("Stretch", "Stretch @ Font Details Sidebar Property", "Property title for `font-stretch`."));
+                break;
```
Comment (rcaliman-apple) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:333]: Please don't update the spacing of unchanged code. It makes it difficult over time to track in which commit a line was added or significantly changed when running something like `git blame`.

Ditto for other places in this PR.

**Review by rcaliman-apple — CHANGES_REQUESTED** [review]
This task can't be solved exclusively on the client-side. You'll need to make changes to the backend to identify, collect, and expose the information about fonts loaded by the inspected page itself.

**Comment by francescorn** [comment]
See for the current pr: https://github.com/WebKit/WebKit/pull/25500
